### PR TITLE
Fix #575 - Add extended set validation

### LINKED ISSUE #575: Extended RPSL set name validation and authentication
Creation of RPSL set objects, as-set, filter-set, peering-set, route-set and rtr-set, currently has few authentication restrictions. If a user has access to any mntner, they can create new sets. For as-set there is a requirement for the first element to be an AS number, e.g. “AS65537:AS-EXAMPLE” when creating new objects as of IRRD 4.2, though this can be disabled.

The plan for this issue is to extend IRRD with new possible requirements for each set type:

 1. All newly created set names will have to follow the hierarchical naming scheme where the first element is an AS number.
 2. Whenever creating a new set object, the database must have an aut-num object for the AS number in the first element.
 3. Authentication must pass for that aut-num (at least one valid authentication method on at least one mnt-by of the aut-num), in addition to normal object creation authentication requirements.

Per set type (as-set, filter-set, etc.) a setting will be added for one of the following:

 1. Prefix not required, no prefix validation.
 2. Prefix not required, but if prefix is present is present opportunistic validation (the prefix may be missing, the aut-num may not yet exist, but if there is a prefix and the aut-num does exist, IRRD must validate).
 3. Prefix required, 'weak' / opportunistic validation (the aut-num may not yet
exist, but if it does IRRD must validate).
 4. Required, strict validation (the aut-num must exist and validate).

In all settings, existing objects can still be modified and deleted, because they may be referenced from other objects from RADB or other IRRs. Forcing a name change to modify existing objects would therefore be a high burden on users.
When modifying or deleting existing set objects, authentication only needs to pass on at least one method on at least one mnt-by on the set object itself, i.e. aut-num authorisation is only required when creating a set object.

### PR DESCRIPTION
**Pull request by mxsasha**
docs:
https://irrd.readthedocs.io/en/extended-set-575-2/admins/configuration/#authentication-and-validation
https://irrd.readthedocs.io/en/extended-set-575-2/users/database-changes/#related-maintainers-in-set-objects

- [x] fix inline todo's
- [x] authenticate_related_mntners setting may now be confusing
- [x] write docs
- [x] decide on appropriate defaults for new auth.set_creation setting
- [x] Look into upgrade path: when people migrate with 4.1 settings, should we still accept those? Concerns one rename and one removed setting. Keep an eye on confusing/conflicting settings and defaults.
- [x] Write release notes for the upgrade path.
- [x] review/refactor all code/docs
- [x] deploy and verify working on test instance
- [x] cleanup test_validators.py
- [x] check integration test

### CODE REVIEW

**Comment by Maxnoue** [comment]
Thank you ..

**Review by Maxnoue — COMMENTED** [review]
Thank you

File: irrd/conf/__init__.py [hunk 1]
```diff
@@ -289,18 +212,24 @@ def _check_staging_config(self) -> List[str]:
...
         config = self.user_config_staging
 
-        for key, value in config.items():
-            if key in ['sources', 'access_lists']:
-                continue
-            known = KNOWN_CONFIG_KEYS.get(key)
-            if known is None:
-                errors.append(f'Unknown setting key: {key}')
+        def _validate_subconfig(key, value):
             if hasattr(value, 'items'):
```
Comment (mxsasha) [inline irrd/conf/__init__.py]: Maybe this should be a dict check?

File: irrd/conf/__init__.py [hunk 2]
```diff
@@ -289,18 +212,24 @@ def _check_staging_config(self) -> List[str]:
...
+                    known_sub = self.known_config_keys.get(subkey)
+
                     if known_sub is None:
                         errors.append(f'Unknown setting key: {subkey}')
+                    _validate_subconfig(subkey, value2)
+
+        for key, value in config.items():
+            if key in ['sources', 'access_lists']:
+                continue
+            known = self.known_config_keys.get(key)
```
Comment (mxsasha) [inline irrd/conf/__init__.py]: Why is this variable created? Seems obsolete

File: docs/users/database-changes.rst [hunk 3]
```diff
@@ -282,7 +284,32 @@ order, only considering the first match:
...
+The details of this behaviour and the strictness of the checks are
+:ref:`configured by the IRR operator <conf-auth-set-creation>`. This may
+include a requirement to:
+
+* Include an ASN prefix in the name of your set, e.g. ``AS65537:AS-EXAMPLE``
+  being valid, but ``AS-EXAMPLE`` being invalid.
+* Pass authentication for the corresponding `aut-num`, e.g. AS65537 in the
+  last example, skipping this check if the `aut-num` does not exist.
+* Pass authentication for the corresponding `aut-num`, e.g. AS65537 in the
+  last example, failing this check if the `aut-num` does not exist.
```
Comment (mxsasha) [inline docs/users/database-changes.rst]:
```suggestion
  example, failing this check if the `aut-num` does not exist.
```

File: docs/users/database-changes.rst [hunk 4]
```diff
@@ -282,7 +284,32 @@ order, only considering the first match:
...
+`rtr-set`.
+
+The details of this behaviour and the strictness of the checks are
+:ref:`configured by the IRR operator <conf-auth-set-creation>`. This may
+include a requirement to:
+
+* Include an ASN prefix in the name of your set, e.g. ``AS65537:AS-EXAMPLE``
+  being valid, but ``AS-EXAMPLE`` being invalid.
+* Pass authentication for the corresponding `aut-num`, e.g. AS65537 in the
+  last example, skipping this check if the `aut-num` does not exist.
```
Comment (mxsasha) [inline docs/users/database-changes.rst]:
```suggestion
  example, skipping this check if the `aut-num` does not exist.
```

File: docs/releases/4.3.0.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+In version 4.2, IRRd required newly created authoritative `as-set` objects
+to have a hierarchical name where the first element is an AS number.
+In 4.3, this feature has been significantly expanded.
+
+For all RPSL set objects, IRRd can now be configured to require:
+
+* Including an ASN prefix in the name of the set, e.g. ``AS65537:AS-EXAMPLE``
+  being valid, but ``AS-EXAMPLE`` being invalid.
+* Passing authentication for the corresponding `aut-num`, e.g. AS65537 in the
+  last example, skipping this check if the `aut-num` does not exist.
```
Comment (mxsasha) [inline docs/releases/4.3.0.rst]:
```suggestion
  example, skipping this check if the `aut-num` does not exist.
```

File: docs/releases/4.3.0.rst [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+* Passing authentication for the corresponding `aut-num`, e.g. AS65537 in the
+  last example, failing this check if the `aut-num` does not exist.
```
Comment (mxsasha) [inline docs/releases/4.3.0.rst]:
```suggestion
  example, failing this check if the `aut-num` does not exist.
```

File: irrd/updates/validators.py [hunk 7]
```diff
@@ -301,22 +300,30 @@ def _find_related_mntners(self, rpsl_obj_new: RPSLObject) -> Optional[Tuple[str,
         - PK of the related object
         - List of maintainers for the related object (at least one must pass)
         Returns None of no related objects were found that should be authenticated.
```
Comment (mxsasha) [inline irrd/updates/validators.py]:
```suggestion
        Returns None if no related objects were found that should be authenticated.
```

**Pull request merged**